File: src/data/courses/dsa/arrays-and-hashing/array-manipulation/pascals-triangle/code/python/pascals_triangle.py
# ============================================================
# Pascal's Triangle
# ============================================================

from typing import List

# ============================================================
# Solution: Row-by-Row Construction (Optimal)
# Time: O(N²) | Space: O(1) extra
# ============================================================
class Solution:
    def generate(self, numRows: int) -> List[List[int]]:
        result = []

        for i in range(numRows):
            # Start with all 1s
            row = [1] * (i + 1)

            # Fill interior: sum of two elements above
            for j in range(1, i):
                row[j] = result[i - 1][j - 1] + result[i - 1][j]

            result.append(row)

        return result

File: src/data/courses/dsa/arrays-and-hashing/array-manipulation/pascals-triangle/code/java/PascalsTriangle.java
// ============================================================
// Pascal's Triangle
// ============================================================

import java.util.*;

// ============================================================
// Solution: Row-by-Row Construction (Optimal)
// Time: O(N²) | Space: O(1) extra
// ============================================================
class Solution {
    public List<List<Integer>> generate(int numRows) {
        List<List<Integer>> result = new ArrayList<>();

        for (int i = 0; i < numRows; i++) {
            List<Integer> row = new ArrayList<>(Collections.nCopies(i + 1, 1));

            // Fill interior elements
            for (int j = 1; j < i; j++) {
                row.set(j, result.get(i - 1).get(j - 1) + result.get(i - 1).get(j));
            }

            result.add(row);
        }

        return result;
    }
}
